index.js

```diff
@@ -16,58 +14,153 @@ const serializeError = FJS({
-          .header('Retry-After', timeWindow)
+          .header('Retry-After', params.globalTimeWindow)
           .send({
             statusCode: 429,
             error: 'Too Many Requests',
-            message: `Rate limit exceeded, retry in ${after}`
+            message: `Rate limit exceeded, retry in ${params.after}`
           })
       }
     }
   }
+}
+
+function rateLimitPlugin (fastify, settings, next) {
+  if (settings.whiteListInRedis && !settings.redis) {
+    next('you must set redis parameter to use it')
+  }
+
+  const globalParams = {
+    global: settings.global || true
+  }
 
+  globalParams.max = (typeof settings.max === 'number' || typeof settings.max === 'function')
+    ? settings.max
+    : 1000
+
+  globalParams.globalTimeWindow = typeof settings.timeWindow === 'string'
+    ? ms(settings.timeWindow)
+    : typeof settings.timeWindow === 'number'
+      ? settings.timeWindow
+      : 1000 * 60
+
+  const pluginComponent = {
+    app: settings.appName || 'default'
+  }
+
+  globalParams.skipOnError = settings.skipOnError === true
+
+  if (settings.redis) {
+    pluginComponent.store = new RedisStore(settings.redis)
+    if (settings.whiteListInRedis) {
+      pluginComponent.store.addWhiteList(`${pluginComponent.app}:wlg`, settings.whitelist || [], (err) => {
+        if (err && globalParams.skipOnError === false) return next(err)
+      })
+      pluginComponent.isRedis = true
+      pluginComponent.whiteListInRedis = true
+    } else {
+      pluginComponent.whitelist = {
+        global: settings.whitelist || [],
+        endpoint: []
+      }
+      pluginComponent.isRedis = true
+      pluginComponent.whiteListInRedis = false
+    }
+  } else {
+    pluginComponent.store = new LocalStore(settings.cache)
+    pluginComponent.isRedis = false
+    pluginComponent.whiteListInRedis = false
+    pluginComponent.whitelist = {
+      global: settings.whitelist || [],
+      endpoint: {}
+    }
+  }
+
+  globalParams.after = ms(globalParams.globalTimeWindow, { long: true })
+
+  globalParams.keyGenerator = typeof settings.keyGenerator === 'function'
+    ? settings.keyGenerator
+    : (req) => req.raw.ip
+
+  // const makeParams = (p) => { return { ...globalParams, ...p } }
+
+  const makeParams = (p) => { return Object.assign({}, globalParams, p) }
+
+  fastify.addHook('onRoute', (routeOptions) => {
+    if (globalParams.global) {
+      if (routeOptions.config && routeOptions.config.rateLimit && typeof routeOptions.config.rateLimit === 'object') {
+        buildRouteRate(pluginComponent, makeParams(routeOptions.config.rateLimit), routeOptions, next)
+      } else {
+        buildRouteRate(pluginComponent, globalParams, routeOptions, next)
+      }
+    } else {
+      if (routeOptions.config && routeOptions.config.rateLimit && typeof routeOptions.config.rateLimit === 'object') {
+        buildRouteRate(pluginComponent, makeParams(routeOptions.config.rateLimit), routeOptions, next)
```

Comment:
You should not pass `next` here. It's the outer next, and you have already called it when this happens.

Reply:
Yes  I had a doubt about that, I will change that.

Reply:
Do you think it could work to return a new `plugin` from within the `rate-limit-plugin`, like a plugin generated on the fly ? 

What I mean is that I could do something like :   

```
const fp = require('fastify-plugin')
............................
.... some code ....
............................
return fp(buildRouteRate(fastify, {
       components : pluginComponent
       params : , makeParams(routeOptions.config.rateLimit),
       route : routeOptions
}, next))
``` 

instead of : 
```
buildRouteRate(pluginComponent, makeParams(routeOptions.config.rateLimit), routeOptions)

```

Would it be suitable, or it's not really in the logic of the `fp` package ?

Reply:
It won't work. But I don't understand why you'll need this. Also, why are you passing `next` there.

Reply:
the `next` in `buildRouteRate` is a mistake on my side, as you said earlier it's not needed .

I will commit the update at lunch, Sorry.

Reply:
updated.